// @alicloud/rpc-client/dist/client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
// This file is auto-generated, don't edit it
const tea_util_1 = __importDefault(require("@alicloud/tea-util"));
const credentials_1 = __importStar(require("@alicloud/credentials")), $Credential = credentials_1;
const rpc_util_1 = __importDefault(require("@alicloud/rpc-util"));
const $tea = __importStar(require("@alicloud/tea-typescript"));
class Config extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            accessKeyId: 'accessKeyId',
            accessKeySecret: 'accessKeySecret',
            network: 'network',
            suffix: 'suffix',
            type: 'type',
            securityToken: 'securityToken',
            endpoint: 'endpoint',
            protocol: 'protocol',
            regionId: 'regionId',
            productId: 'productId',
            userAgent: 'userAgent',
            readTimeout: 'readTimeout',
            connectTimeout: 'connectTimeout',
            httpProxy: 'httpProxy',
            httpsProxy: 'httpsProxy',
            noProxy: 'noProxy',
            socks5Proxy: 'socks5Proxy',
            socks5NetWork: 'socks5NetWork',
            maxIdleConns: 'maxIdleConns',
            endpointType: 'endpointType',
            openPlatformEndpoint: 'openPlatformEndpoint',
        };
    }
    static types() {
        return {
            accessKeyId: 'string',
            accessKeySecret: 'string',
            network: 'string',
            suffix: 'string',
            type: 'string',
            securityToken: 'string',
            endpoint: 'string',
            protocol: 'string',
            regionId: 'string',
            productId: 'string',
            userAgent: 'string',
            readTimeout: 'number',
            connectTimeout: 'number',
            httpProxy: 'string',
            httpsProxy: 'string',
            noProxy: 'string',
            socks5Proxy: 'string',
            socks5NetWork: 'string',
            maxIdleConns: 'number',
            endpointType: 'string',
            openPlatformEndpoint: 'string',
        };
    }
}
exports.Config = Config;
class Client {
    constructor(config) {
        let credentialConfig = null;
        if (tea_util_1.default.isUnset($tea.toMap(config))) {
            config = new Config({});
            this._credential = new credentials_1.default(null);
        }
        else {
            credentialConfig = new $Credential.Config({
                accessKeyId: config.accessKeyId,
                type: config.type,
                accessKeySecret: config.accessKeySecret,
                securityToken: config.securityToken,
            });
            this._credential = new credentials_1.default(credentialConfig);
        }
        this._network = config.network;
        this._suffix = config.suffix;
        this._endpoint = config.endpoint;
        this._protocol = config.protocol;
        this._regionId = config.regionId;
        this._userAgent = config.userAgent;
        this._readTimeout = config.readTimeout;
        this._connectTimeout = config.connectTimeout;
        this._httpProxy = config.httpProxy;
        this._httpsProxy = config.httpsProxy;
        this._noProxy = config.noProxy;
        this._socks5Proxy = config.socks5Proxy;
        this._socks5NetWork = config.socks5NetWork;
        this._maxIdleConns = config.maxIdleConns;
        this._endpointType = config.endpointType;
        this._openPlatformEndpoint = config.openPlatformEndpoint;
    }
    doRequest(action, protocol, method, version, authType, query, body, runtime) {
        return __awaiter(this, void 0, void 0, function* () {
            let _runtime = {
                timeouted: "retry",
                readTimeout: tea_util_1.default.defaultNumber(runtime.readTimeout, this._readTimeout),
                connectTimeout: tea_util_1.default.defaultNumber(runtime.connectTimeout, this._connectTimeout),
                httpProxy: tea_util_1.default.defaultString(runtime.httpProxy, this._httpProxy),
                httpsProxy: tea_util_1.default.defaultString(runtime.httpsProxy, this._httpsProxy),
                noProxy: tea_util_1.default.defaultString(runtime.noProxy, this._noProxy),
                maxIdleConns: tea_util_1.default.defaultNumber(runtime.maxIdleConns, this._maxIdleConns),
                retry: {
                    retryable: runtime.autoretry,
                    maxAttempts: tea_util_1.default.defaultNumber(runtime.maxAttempts, 3),
                },
                backoff: {
                    policy: tea_util_1.default.defaultString(runtime.backoffPolicy, "no"),
                    period: tea_util_1.default.defaultNumber(runtime.backoffPeriod, 1),
                },
                ignoreSSL: runtime.ignoreSSL,
            };
            let _lastRequest = null;
            let _now = Date.now();
            let _retryTimes = 0;
            while ($tea.allowRetry(_runtime['retry'], _retryTimes, _now)) {
                if (_retryTimes > 0) {
                    let _backoffTime = $tea.getBackoffTime(_runtime['backoff'], _retryTimes);
                    if (_backoffTime > 0) {
                        yield $tea.sleep(_backoffTime);
                    }
                }
                _retryTimes = _retryTimes + 1;
                try {
                    let request_ = new $tea.Request();
                    request_.protocol = tea_util_1.default.defaultString(this._protocol, protocol);
                    request_.method = method;
                    request_.pathname = "/";
                    request_.query = rpc_util_1.default.query(Object.assign({ Action: action, Format: "json", Timestamp: rpc_util_1.default.getTimestamp(), Version: version, SignatureNonce: tea_util_1.default.getNonce() }, query));
                    if (!tea_util_1.default.isUnset(body)) {
                        let tmp = tea_util_1.default.anyifyMapValue(rpc_util_1.default.query(body));
                        request_.body = new $tea.BytesReadable(tea_util_1.default.toFormString(tmp));
                    }
                    // endpoint is setted in product client
                    request_.headers = {
                        host: this._endpoint,
                        'user-agent': this.getUserAgent(),
                    };
                    if (!tea_util_1.default.equalString(authType, "Anonymous")) {
                        let accessKeyId = yield this.getAccessKeyId();
                        let accessKeySecret = yield this.getAccessKeySecret();
                        let securityToken = yield this.getSecurityToken();
                        if (!tea_util_1.default.empty(securityToken)) {
                            request_.query["SecurityToken"] = securityToken;
                        }
                        request_.query["SignatureMethod"] = "HMAC-SHA1";
                        request_.query["SignatureVersion"] = "1.0";
                        request_.query["AccessKeyId"] = accessKeyId;
                        let signedParam = Object.assign(Object.assign({}, request_.query), rpc_util_1.default.query(body));
                        request_.query["Signature"] = rpc_util_1.default.getSignatureV1(signedParam, request_.method, accessKeySecret);
                    }
                    _lastRequest = request_;
                    let response_ = yield $tea.doAction(request_, _runtime);
                    let obj = yield tea_util_1.default.readAsJSON(response_.body);
                    let res = tea_util_1.default.assertAsMap(obj);
                    if (tea_util_1.default.is4xx(response_.statusCode) || tea_util_1.default.is5xx(response_.statusCode)) {
                        throw $tea.newError({
                            message: res["Message"],
                            data: res,
                            code: res["Code"],
                        });
                    }
                    return res;
                }
                catch (ex) {
                    if ($tea.isRetryable(ex)) {
                        continue;
                    }
                    throw ex;
                }
            }
            throw $tea.newUnretryableError(_lastRequest);
        });
    }
    getUserAgent() {
        let userAgent = tea_util_1.default.getUserAgent(this._userAgent);
        return userAgent;
    }
    getAccessKeyId() {
        return __awaiter(this, void 0, void 0, function* () {
            if (tea_util_1.default.isUnset(this._credential)) {
                return "";
            }
            let accessKeyId = yield this._credential.getAccessKeyId();
            return accessKeyId;
        });
    }
    getAccessKeySecret() {
        return __awaiter(this, void 0, void 0, function* () {
            if (tea_util_1.default.isUnset(this._credential)) {
                return "";
            }
            let secret = yield this._credential.getAccessKeySecret();
            return secret;
        });
    }
    getSecurityToken() {
        return __awaiter(this, void 0, void 0, function* () {
            if (tea_util_1.default.isUnset(this._credential)) {
                return "";
            }
            let token = yield this._credential.getSecurityToken();
            return token;
        });
    }
    checkConfig(config) {
        if (tea_util_1.default.empty(this._endpointRule) && tea_util_1.default.empty(config.endpoint)) {
            throw $tea.newError({
                name: "ParameterMissing",
                message: "'config.endpoint' can not be empty",
            });
        }
    }
}
exports.default = Client;
